app/browser/api/ledger.js

```diff
@@ -0,0 +1,2187 @@
+    (err, provider, result) => {
+      if (err) {
+        return console.warn('ledger balance warning: ' + JSON.stringify(err, null, 2))
+      }
+      appActions.onLedgerBalanceReceived(result.unconfirmed)
+    })
+}
+
+const balanceReceived = (state, unconfirmed) => {
+  if (typeof unconfirmed === 'undefined') {
+    return state
+  }
+
+  if (unconfirmed > 0) {
+    const result = (unconfirmed / 1e8).toFixed(4)
+    if (ledgerState.getInfoProp(state, 'unconfirmed') === result) {
+      return state
+    }
+
+    state = ledgerState.setInfoProp(state, 'unconfirmed', result)
+    if (clientOptions.verboseP) {
+      console.log('\ngetBalance refreshes ledger info: ' + ledgerState.getInfoProp(state, 'unconfirmed'))
+    }
+    return updateLedgerInfo(state)
+  }
+
+  if (ledgerState.getInfoProp(state, 'unconfirmed') === '0.0000') {
+    return state
+  }
+
+  if (clientOptions.verboseP) console.log('\ngetBalance refreshes payment info')
+  return getPaymentInfo(state)
+}
+
+const callback = (err, result, delayTime) => {
+  if (clientOptions.verboseP) {
+    console.log('\nledger client callback: clientP=' + (!!client) + ' errP=' + (!!err) + ' resultP=' + (!!result) +
+      ' delayTime=' + delayTime)
+  }
+
+  if (err) {
+    console.log('ledger client error(1): ' + JSON.stringify(err, null, 2) + (err.stack ? ('\n' + err.stack) : ''))
+    if (!client) return
+
+    if (typeof delayTime === 'undefined') {
+      delayTime = random.randomInt({min: miliseconds.minute, max: 10 * miliseconds.minute})
+    }
+  }
+
+  appActions.onLedgerCallback(result, delayTime)
+}
+
+const onCallback = (state, result, delayTime) => {
+  let results
+  let entries = client && client.report()
+
+  if (!result) {
+    return run(state, delayTime)
+  }
+
+  const regularResults = result.toJS()
+
+  if (client && result.getIn(['properties', 'wallet'])) {
+    if (!ledgerState.getInfoProp(state, 'created')) {
+      setPaymentInfo(getSetting(settings.PAYMENTS_CONTRIBUTION_AMOUNT))
+    }
+
+    state = getStateInfo(state, regularResults) // TODO optimize if possible
+    state = getPaymentInfo(state)
+  }
+
+  state = cacheRuleSet(state, regularResults.ruleset)
+  if (result.has('rulesetV2')) {
+    results = regularResults.rulesetV2 // TODO optimize if possible
+    result = result.delete('rulesetV2')
+
+    entries = []
+    results.forEach((entry) => {
+      const key = entry.facet + ':' + entry.publisher
+
+      if (entry.exclude !== false) {
+        entries.push({type: 'put', key: key, value: JSON.stringify(underscore.omit(entry, ['facet', 'publisher']))})
+      } else {
+        entries.push({type: 'del', key: key})
+      }
+    })
+
+    v2RulesetDB.batch(entries, (err) => {
+      if (err) return console.error(v2RulesetPath + ' error: ' + JSON.stringify(err, null, 2))
+
+      if (entries.length === 0) return
+
+      const publishers = ledgerState.getPublishers(state)
+      for (let item of publishers) {
```

Comment:
for..of 😢  (would benefit from instead being the index based, like `for(let i=0;i<publishers.length;i++)`)

Reply:
answered above